File: core/api/typescript/src/lib/services/service_context.ts
import { err, ok, Result } from 'neverthrow';
import { newExecCommandArgs } from '../constructor_calls';
import type { ExecCommandArgs } from '../../kurtosis_core_rpc_api_bindings/api_container_service_pb';
import type { PortSpec } from './port_spec';
import type { ServiceID } from './service';
import type { SharedPath } from './shared_path';
import { GenericApiContainerClient } from '../enclaves/generic_api_container_client';

// Docs available at https://docs.kurtosistech.com/kurtosis-core/lib-documentation
export class ServiceContext {

    private readonly client: GenericApiContainerClient
    private readonly serviceId: ServiceID
    private readonly sharedDirectory: SharedPath
    private readonly privateIpAddress: string
    private readonly privatePorts: Map<string, PortSpec>
    private readonly publicIpAddress: string
    private readonly publicPorts: Map<string, PortSpec>

    constructor(
        client: GenericApiContainerClient,
        serviceId: ServiceID,
        sharedDirectory: SharedPath,
        privateIpAddress: string,
        privatePorts: Map<string, PortSpec>,
        publicIpAddress: string,
        publicPorts: Map<string, PortSpec>
        ){

        this.client = client
        this.serviceId = serviceId
        this.sharedDirectory = sharedDirectory
        this.privateIpAddress = privateIpAddress
        this.privatePorts = privatePorts
        this.publicIpAddress = publicIpAddress
        this.publicPorts = publicPorts
    }

    // Docs available at https://docs.kurtosistech.com/kurtosis-core/lib-documentation
    public getServiceID(): ServiceID { 
        return this.serviceId;
    }

    // Docs available at https://docs.kurtosistech.com/kurtosis-core/lib-documentation
    public getSharedDirectory(): SharedPath {
        return  this.sharedDirectory
    }

    // Docs available at https://docs.kurtosistech.com/kurtosis-core/lib-documentation
    public getPrivateIPAddress(): string {
        return this.privateIpAddress
    }

    // Docs available at https://docs.kurtosistech.com/kurtosis-core/lib-documentation
    public getPrivatePorts(): Map<string, PortSpec> {
        return this.privatePorts
    }

    // Docs available at https://docs.kurtosistech.com/kurtosis-core/lib-documentation
    public getMaybePublicIPAddress(): string {
        return this.publicIpAddress
    }

    // Docs available at https://docs.kurtosistech.com/kurtosis-core/lib-documentation
    public getPublicPorts(): Map<string, PortSpec> {
        return this.publicPorts
    }

    // Docs available at https://docs.kurtosistech.com/kurtosis-core/lib-documentation
    public async execCommand(command: string[], ): Promise<Result<[number, string], Error>> {
        const execCommandArgs: ExecCommandArgs = newExecCommandArgs(this.serviceId, command);

        const execCommandResponseResult = await this.client.execCommand(execCommandArgs)
        if(execCommandResponseResult.isErr()){
            return err(execCommandResponseResult.error)
        }

        const execCommandResponse = execCommandResponseResult.value
        return ok([execCommandResponse.getExitCode(), execCommandResponse.getLogOutput()]);
    }
}

File: core/api/golang/lib/services/service_context.go
/*
 *    Copyright 2021 Kurtosis Technologies Inc.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 *
 */

package services

import (
	"context"
	"github.com/kurtosis-tech/kurtosis-client/golang/kurtosis_core_rpc_api_bindings"
	"github.com/kurtosis-tech/kurtosis-client/golang/lib/binding_constructors"
	"github.com/kurtosis-tech/stacktrace"
)

// Docs available at https://docs.kurtosistech.com/kurtosis-client/lib-documentation
type ServiceContext struct {
	client			kurtosis_core_rpc_api_bindings.ApiContainerServiceClient
	serviceId		ServiceID
	ipAddress		string
	sharedDirectory	*SharedPath
}

func NewServiceContext(
		client kurtosis_core_rpc_api_bindings.ApiContainerServiceClient,
		serviceId ServiceID,
		ipAddress string,
	    sharedDirectory *SharedPath) *ServiceContext {
	return &ServiceContext{
		client:				client,
		serviceId:			serviceId,
		ipAddress:			ipAddress,
		sharedDirectory:	sharedDirectory,
	}
}

// Docs available at https://docs.kurtosistech.com/kurtosis-client/lib-documentation
func (self *ServiceContext) GetServiceID() ServiceID {
	return self.serviceId
}

// Docs available at https://docs.kurtosistech.com/kurtosis-client/lib-documentation
func (self *ServiceContext) GetIPAddress() string {
	return self.ipAddress
}

// Docs available at https://docs.kurtosistech.com/kurtosis-client/lib-documentation
func (self *ServiceContext) GetSharedDirectory() *SharedPath {
	return self.sharedDirectory
}

// Docs available at https://docs.kurtosistech.com/kurtosis-client/lib-documentation
func (self *ServiceContext) ExecCommand(command []string) (int32, string, error) {
	serviceId := self.serviceId
	args := binding_constructors.NewExecCommandArgs(string(serviceId), command)
	resp, err := self.client.ExecCommand(context.Background(), args)
	if err != nil {
		return 0, "", stacktrace.Propagate(
			err,
			"An error occurred executing command '%v' on service '%v'",
			command,
			serviceId)
	}
	return resp.ExitCode, resp.LogOutput, nil
}
